Libplanet/Blockchain/BlockChain.cs

```diff
@@ -682,63 +685,120 @@ public long GetNextTxNonce(Address address)
+        /// Zero and negative values are treated as 1. If it is omitted or more than
+        /// <see cref="Policy"/>.<see cref="IBlockPolicy{T}.MaxTransactionsPerBlock"/>, it will be
+        /// treated as <see cref="Policy"/>.<see cref="IBlockPolicy{T}.MaxTransactionsPerBlock"/>.
+        /// </param>
         /// <param name="cancellationToken">
         /// A cancellation token used to propagate notification that this
         /// operation should be canceled.
         /// </param>
         /// <returns>An awaitable task with a <see cref="Block{T}"/> that is mined.</returns>
         /// <exception cref="OperationCanceledException">Thrown when
         /// <see cref="BlockChain{T}.Tip"/> is changed while mining.</exception>
-        /// <exception cref="ArgumentOutOfRangeException">Thrown when
-        /// <paramref name="txBatchSize"/> is negative.</exception>
         public async Task<Block<T>> MineBlock(
             Address miner,
             DateTimeOffset currentTime,
             bool append = true,
-            int txBatchSize = int.MaxValue,
+            int? maxTransactions = null,
             CancellationToken cancellationToken = default(CancellationToken)
         )
         {
-            if (txBatchSize < 0)
-            {
-                throw new ArgumentOutOfRangeException(nameof(txBatchSize));
-            }
+            maxTransactions = Math.Max(
+                Math.Min(
+                    maxTransactions ?? Policy.MaxTransactionsPerBlock,
+                    Policy.MaxTransactionsPerBlock
+                ),
+                1
+            );
 
             long index = Store.CountIndex(Id);
             long difficulty = Policy.GetNextBlockDifficulty(this);
             HashDigest<SHA256>? prevHash = Store.IndexBlockHash(Id, index - 1);
 
-            IEnumerable<Transaction<T>> stagedTransactions = Store
+            Transaction<T>[] stagedTransactions = Store
                 .IterateStagedTransactionIds()
                 .Select(Store.GetTransaction<T>)
                 .GroupBy(tx => tx.Signer)
                 .SelectMany(grp => grp.Select((tx, idx) => (tx, idx)))
                 .OrderBy(t => t.Item2)
-                .Select(t => t.Item1);
+                .Select(t => t.Item1)
+                .ToArray();
 
             var transactionsToMine = new List<Transaction<T>>();
 
+            // Makes an empty block to estimate the length of bytes without transactions.
+            var estimatedBytes = new Block<T>(
+                index: index,
+                difficulty: difficulty,
+                totalDifficulty: Tip.TotalDifficulty,
+                nonce: default,
+                miner: miner,
+                previousHash: prevHash,
+                timestamp: currentTime,
+                transactions: new Transaction<T>[0]
+            ).BytesLength;
+            int maxBlockBytes = Math.Max(Policy.GetMaxBlockBytes(index), 1);
```

Comment:
What is the meaning of `1`?

Reply:
This makes the `maxBlockBytes` 1 at least even if `Policy.GetMaxBlockBytes()` is less than 1 (e.g., `Max(-1, 1)` → `1`, `Max(5, 1)` → `5`).

Reply:
I was wondering why the value is exactly 1 🤔

Reply:
This implements the below documented behavior:

<https://github.com/planetarium/libplanet/blob/e7d963bae4a757f45ac808d679e31a060a5ed780/Libplanet/Blockchain/Policies/IBlockPolicy.cs#L81-L82>

which assumes there is no block taking 0 bytes or negative length of bytes.